src/signer/mod.rs

```diff
@@ -0,0 +1,251 @@
+//! Traits and types for creating signed urls
+//!
+//! Signed URLs are used to provide access to data in object stores without
+//! requiring the client to authenticate with the object store. The client
+//! can access the data using the signed URL, which is valid for a limited
+//! time.
+//!
+//! The Delta Sharing protocol requires signed urls to access data in the object
+//! store. This module provides a trait for creating signed urls and
+//! implementations of this trait for several popular (cloud) object stores.
+
+#![warn(missing_docs)]
+
+use std::time::Duration;
+use std::{fmt, time::SystemTime};
+
+use async_trait::async_trait;
+
+pub mod s3;
+
+/// Interface for creating signed URLs from object store specific URIs.
+#[async_trait]
+pub trait Signer: Send + Sync {
+    /// Create a signed url from a (cloud) object store URI that is valid from
+    /// now and expires after the specified duration.
+    async fn sign(&self, uri: &str, expires_in: Duration) -> Result<SignedUrl, SignerError>;
```

Comment:
I was not aware that kernel would only exposed parsed URLs. In that case it would make sense to take a `url::Url`. A (future) consideration could be to take a `impl IntoUrl` (ref: https://docs.rs/reqwest/latest/reqwest/trait.IntoUrl.html). I'll make the change to `url::Url` for now